src/Microsoft.AspNetCore.Razor.Language/DefaultRazorProjectEngine.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class DefaultRazorProjectEngine : RazorProjectEngine
+    {
+        public DefaultRazorProjectEngine(
+            RazorEngine engine,
+            RazorProjectFileSystem projectFileSystem,
+            IReadOnlyList<IRazorProjectEngineFeature> features)
+        {
+            if (engine == null)
+            {
+                throw new ArgumentNullException(nameof(engine));
+            }
+
+            if (projectFileSystem == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileSystem));
+            }
+
+            if (features == null)
+            {
+                throw new ArgumentNullException(nameof(features));
+            }
+
+            Engine = engine;
+            ProjectFileSystem = projectFileSystem;
+            Features = features;
+
+            for (var i = 0; i < features.Count; i++)
+            {
+                features[i].ProjectEngine = this;
+            }
+        }
+
+        public override RazorProjectFileSystem ProjectFileSystem { get; }
+
+        public override RazorEngine Engine { get; }
+
+        public override IReadOnlyList<IRazorProjectEngineFeature> Features { get; }
+
+        public override RazorCodeDocument Process(string filePath)
+        {
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            var projectItem = ProjectFileSystem.GetItem(filePath);
+            var sourceDocument = RazorSourceDocument.ReadFrom(projectItem);
+            var codeDocument = Process(sourceDocument);
+
+            return codeDocument;
+        }
+
+        public override RazorCodeDocument Process(RazorSourceDocument sourceDocument)
```

Comment:
It's not clear to me that this will be useful in the long term. Who's the consumer of this? The editor?

Is the expectation that the file system reflects the on-disk state? or the in-memory state (IDE)? If it's the latter, is passing in the source document useful?